Remove "Create Workflow" form and allow workflow creation in editor

ahmedhamidawan opened the pull request:
Skip the step of Creating a workflow, and instead go straight into an empty editor, add nodes, tags, connections...; workflow will only be created once you Save and provide the worfklow a name. Fixes https://github.com/galaxyproject/galaxy/issues/16912

https://github.com/galaxyproject/galaxy/assets/78516064/90df5d03-e77f-4434-b3f0-db2a9c5b30b7

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
Remove "Create Workflow" form and allow workflow creation in editor
Skip the step of Creating a workflow, and instead go straight into an
empty editor, add nodes, tags, connections...; workflow will only
be created once you Save and provide the worfklow a name and annotation.
Fixes https://github.com/galaxyproject/galaxy/issues/16912

ElectronicBlueberry commented on client/src/components/Workflow/Editor/Index.vue:
```diff
@@ -191,6 +196,8 @@ import ToolPanel from "@/components/Panels/ToolPanel.vue";
 import FormDefault from "@/components/Workflow/Editor/Forms/FormDefault.vue";
 import FormTool from "@/components/Workflow/Editor/Forms/FormTool.vue";
 
+const NEW_TEMP_ID = "new_temp_workflow";
```
Could we generate this? Maybe using uuid. Reason for this is that in the future we may want to display multiple instances of the workflow editor at once, which is the reason for the store scoping.

ahmedhamidawan replied:
Yes that is a good idea 👍

ahmedhamidawan replied:
~~Created a `uuid` composable that is being used instead now~~ Using the existing `uid` composable instead

martenson commented on lib/galaxy/webapps/galaxy/controllers/workflow.py:
```diff
@@ -510,24 +522,28 @@ def editor(self, trans, id=None, workflow_id=None, version=None, **kwargs):
             "workflows": workflows,
         }
 
+        # if existing workflow, add add its data to the model
```
```suggestion
        # for existing workflow add its data to the model
```

martenson replied:
This has been resolved with the typo still in. Next time you can use the "commit suggestion" button to save yourself some work.

martenson commented on client/src/components/Workflow/Editor/Index.vue:
```diff
@@ -541,6 +555,40 @@ export default {
         },
+        async onCreate() {
+            if (!this.name || !this.annotation) {
+                const response = "Must provide name and annotation before creation...";
```
These are user-facing messages. We (usually) know exactly what happened and specifically here we are asking user to correct. I believe something like `Please provide a name for your workflow.` and avoiding using ellipsis throughout the status messages (to create certainty instead) would be an improvement.

ahmedhamidawan committed:
do not require annotation for wfs

ahmedhamidawan committed:
use `uid` composable for temp wf id...
handle errors caused by wf creation without annotation

ahmedhamidawan committed:
update seleniums to work with new way of creating workflows

ElectronicBlueberry commented on client/src/composables/uuid.ts:
```diff
@@ -0,0 +1,35 @@
+import Vue, { ref } from "vue";
```
I do not like the naming of this composable as it does not return an uuid. It also seems to have large overlap with the existing useUid composable. Would it be possible to use that instead of making a new composable?

dannon replied:
Yep, this should just be the existing `useUid`, it already does exactly what was discussed in wg-uiux.

ahmedhamidawan replied:
Sorry, I somehow completely missed that one, used that one instead ✅

ahmedhamidawan committed:
change `router.push` -> `window.location` to scope wf stores to new id

ahmedhamidawan committed:
selenium add render wait to wait for window location change
also fix jest for wf editor attributes by specifying version prop

ahmedhamidawan committed:
fix integrations caused by using `tag_handler`

ElectronicBlueberry committed:
provide id as ref object instead of string

ElectronicBlueberry committed:
add scope pointers

ElectronicBlueberry committed:
add scope wrapper to all scoped workflow stores

ElectronicBlueberry committed:
use scope pointer to avoid copying store state

ElectronicBlueberry committed:
allow passing a plain string to provideScopedWorkflowStores

ElectronicBlueberry committed:
use unref instead of value to simplify testing

ElectronicBlueberry committed:
use replace to avoid redundant history entry
show save progress dialog on first save
remove modal which is immediately hidden after creation

ElectronicBlueberry committed:
wait for workflow to load before closing loading indicator

ElectronicBlueberry committed:
hide attributes panel while loading

ahmedhamidawan committed:
use `create` endpoint for name, ann & tags, else use save as endpoint

Also, use `scopePointerStore` for actual Save As operation as well

ahmedhamidawan committed:
`scopePointerStore` always checks for existing scope

martenson reviewed (approved):
Works well. I am not sold on forcing the name without providing a default. Especially since I can just subsequently erase the name and still save it (not related to this PR).
If that is still preferred then a minor follow up I'd like to see is to have the missing name "validation modal" replaced with an in-line validation tooltip. And maybe a toast on save?

mvdbeek commented on lib/galaxy/managers/workflows.py:
```diff
@@ -687,13 +687,23 @@ def update_workflow_from_raw_description(
             stored_workflow.latest_workflow = workflow
             workflow.stored_workflow = stored_workflow
 
+        data = raw_workflow_description.as_dict
```
@ahmedhamidawan do you remember why you made that change ?

ahmedhamidawan replied:
The purpose of doing this was to be able to create a workflow with tags as part of the payload (`raw_workflow_description`); instead of first creating a workflow, and then updating it with tags. Therefore, here, I created this `data` dict to check for tags (similar to what we do for name and annotation in the same function later in L701

ahmedhamidawan replied:
Oh wait, this was back when we used the `save_workflow_as` route to create workflows, so this was needed back then, Now we use the fast api POST workflows route, so this change might not be needed anymore as much,,,
